HSEARCH-4563 queryString predicate for advanced, Lucene-syntax query strings / HSEARCH-5073 Add minimum_should_match to simple query string predicate (#3937)
hibernate/hibernate-search: documentation/src/main/asciidoc/public/reference/_search-dsl-predicate.adoc

If there is a need to prevent users from using leading wildcards, `.allowLeadingWildcard(..)` can be called
with a `false` value to disallow such queries.

.Matching a query string: disallowing leading wildcards
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/predicate/PredicateDslIT.java[tags=queryString-allow-leading-wildcard]
----
In this example, a query uses the wildcard at the end of the word; hence, it is acceptable regardless of the option value.
If queries like `?obot` or `*bot` were supplied, and at the same time, leading wildcards were disallowed
it would have resulted in an exception being thrown.
====

Note that this option affects not only the initial (entire) query string but also the substring queries.
For example, in a query string `robot ?etective` the wildcard `?` is not a leading character,
but this query is broken down as two clauses for `robot` and `?etecitve` where in the second
clause, the `?` wildcard becomes a leading character.

[[search-dsl-predicate-query-string-enable-position-increments]]
=== Enabling position increments

> Position increments are enabled by default.
> When disabled, resulting queries are not aware of position increments
> resulted from removing stop words by a `stopwords` filter.
> This, for example, can lead to a change in the behaviour of phrase queries:

Reviewer: ```suggestion
Position increments are enabled by default, allowing phrase queries to take into account stop words removed by a `stopwords` filter.

Position increments can be disabled as shown below,
leading to a change in the behaviour of phrase queries:
```

assuming that there is a phrase `book at the shelve` in the document, and the stop filter removes both `at` and `the`,
with disabled position increments, phrase query `"book shelve"` will not match such document.

.Matching a query string: disabling position increments
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/predicate/PredicateDslIT.java[tags=queryString-enable-position-increments]
----
====

[[search-dsl-predicate-query-string-rewrite-method]]
=== Rewrite method

> The rewrite method determines how the backend's query parser rewrites and scores multi-term queries.
> To change the default `CONSTANT_SCORE` rewrite method, one of the allowed `RewriteMethod` enum values can be passed

Reviewer: ```suggestion
The rewrite method determines how the backend's query parser rewrites and scores multi-term queries.
To change the default `CONSTANT_SCORE` rewrite method, one of the allowed `RewriteMethod` enum values can be passed
```
```suggestion
The rewrite method determines how the backend's query parser rewrites and scores multi-term queries.

To change the default `CONSTANT_SCORE` rewrite method, one of the allowed `RewriteMethod` enum values can be passed
```

> to `.rewriteMethod(RewriteMethod)`/`rewriteMethod(RewriteMethod, int)`.
> We will not go into details about different rewrite methods in this guide.

Reviewer: ```suggestion
to `.rewriteMethod(RewriteMethod)`/`rewriteMethod(RewriteMethod, int)`.

We will not go into details about different rewrite methods in this guide.
```

.Fine-tuning `should` clauses matching requirements with `minimumShouldMatch`
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/predicate/PredicateDslIT.java[tags=simpleQueryString-minimum-should-match]
----
====

This is similar to <<search-dsl-predicate-boolean-minimumshouldmatch,boolean>> or
<<search-dsl-predicate-simple-query-string-minimumshouldmatch,simple query string>>
predicates `minimumShouldMatch` options.

[[search-dsl-predicate-query-string-multiple-fields]]
=== Targeting multiple fields

Optionally, the predicate can target multiple fields.
In that case, the predicate will match documents for which _any_ of the given fields matches.

See <<search-dsl-predicate-common-multiple-fields>>.

IMPORTANT: If targeted fields have different analyzers, an exception will be thrown.
You can avoid this by <<search-dsl-predicate-common-overriding-analysis,picking an analyzer explicitly>>,
but make sure you know what you're doing.

[[search-dsl-predicate-query-string-other]]
=== Other options

* The score of a `queryString` predicate is variable by default,

> but can be <<search-dsl-predicate-common-constantScore,made constant with `.constantScore()`>>.

Reviewer: You said above that the default rewrite method was "CONSTANT_SCORE"... might want to explain that this doesn't mean, in fact, that the score is constant...
